scalableminds/webknossos
# Refactor url hash to json

Refactor the url hash/fragment to alternatively use json format instead of a comma-delimited custom format. I added quite a few tests to make sure that both the csv format as well as the comment links (which also temporarily change the url hash) continue to work. Keep the csv format as the default.

- [x] Certain characters in the fragment string need to be escaped (see https://stackoverflow.com/questions/2849756/list-of-valid-characters-for-the-fragment-identifier-in-an-url). For now, I've used the built-in `encodeURIComponent` to do that. However, the function encodes quite a few characters which don't need to be encoded, resulting in a rather unreadable url hash.  We could write a custom {en,de}coding wrapper which doesn't encode `:`, `.` and `,` to make the hash more readable. On the other hand, we don't really want users to modify the url hash anyways, I'd say. What do you think?

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- Open a url with a hash in the **csv** format and check that it opens correctly and the url state is correctly applied. E.g. append `#10,20,30,0,1.3,2` (`x,y,z,mode,zoomStep,activeNode`)
  - Modify values in the url hash json format
    - Check that the state is applied after pressing Enter
- Open a url with a hash in the **json** format and check that it opens correctly and the url state is correctly applied. E.g. append `#%7B"position"%3A%5B445%2C358%2C501%5D%2C"mode"%3A"orthogonal"%2C"zoomStep"%3A2%2C"activeNode"%3A2%7D`
  - Modify values in the url hash json format
    - Check that an error is thrown if the json format is invalid
- Check that the url hash works in flight mode or when there is no active node
- Create comments with links, for example `jump to node with id #2` and `jump to position #(1,2,3)`. Then click on the links and check that the links work as expected.

### Issues:
- fixes #4596 

------
